### DISCUSSION TITLE
Using Magic Link for auth flow

### DISCUSSION THREAD

**Discussion by moocowdrinksmilk** [AfterClass-io/afterclass.io-v2#26] [Ideas & Feature Requests]
Magic links could provide a better UX for logging in compared to traditional auth that is being implemented now. It also allows for checking of email to ensure the user is a student from SMU. 

Not sure if there are additional costs to enable this, or if its within the budget.

https://supabase.com/docs/guides/auth/auth-magic-link

**Comment by davidlhw** [comment]
Thanks for raising this, it's a great suggestion!

Agree that magic links seem to be a nice alternative to the traditional [`signInWithPassword`](https://supabase.com/docs/reference/javascript/auth-signinwithpassword) authentication for better UX. Another cool thing to note from the [`signInWithOtp` Supabase API Doc](https://supabase.com/docs/reference/javascript/auth-signinwithotp) is that:

> If the user doesn't exist, signInWithOtp() will signup the user instead. 

Which is rather convenient vs the traditional [`signInWithPassword`](https://supabase.com/docs/reference/javascript/auth-signinwithpassword), as `signInWithPassword` would only work if user is already registered _(and has confirmed their email if that option is enabled)_

In terms of cost, right off the top of my head, it doesn't seem to be too much different from the traditional way as new users will need to verify their control of the emails anyway; meaning that we'll need to send 1 email for both scenarios either way.

---

Some additional context on our AuthN requirements - it should ideally:

1. Validate user's credentials in order to access protected resource
2. Verify that user is a student from a supported school by verifying their `.edu` email address

Meaning that something like using [Google Auth Provider with NextAuth](https://next-auth.js.org/providers/google) won't cut it since we can't guarantee that every university has google suite tied to the email like SMU's

---

Other relevant issues: #2 , #4

**Comment by Aztriltus** [comment]
IMO Magic login is not very fast as it takes about 30s in total. If we're also not using oauth in order to check for the student email address, then we're also not taking full advantage of Magic too. Not sure if it's worth the UX trade off. Can't vouch for BE though

**Reply by davidlhw** [reply]
Good point! magic link  might not be ideal for the time being as 30s is definitely too long in terms of UX especially since our average user session duration is only about 1-2 mins.

Magic links might still be nice as an alternative for users who prefers passwordless login.
